File: fusion/normalize.py
# -*- coding: utf-8 -*-
"""Text normalization, ported verbatim from eval/controlled_eval/scoring.py.

Normalization is deliberately shallow: NFD, strip combining marks, lowercase, and
split on `\\w+`. It does NOT fold final sigma, so `ς` and `σ` are distinct tokens.

Do not "improve" this. Every frozen WER in CONTRACT.md was measured through it.

On top of the ported scorer this module adds the raw <-> norm map the fusion
pipeline needs: one raw provider word yields 0..n normalized tokens, and the
fused output has to name the raw word each chosen token came from.
"""
from __future__ import annotations

import re
import unicodedata

NORMALIZER_REV = "wtoks/1"


def norm(s):
    s = unicodedata.normalize("NFD", s or "")
    return "".join(c for c in s if unicodedata.category(c) != "Mn").lower()


def wtoks(s):
    return re.findall(r"\w+", norm(s))


def edist(a, b):
    """Levenshtein distance over any two sequences."""
    n, m = len(a), len(b)
    if n == 0:
        return m
    prev = list(range(m + 1))
    for i in range(1, n + 1):
        cur = [i] + [0] * m
        for j in range(1, m + 1):
            cur[j] = min(prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + (a[i - 1] != b[j - 1]))
        prev = cur
    return prev[m]


def sdi(ref: str, hyp: str) -> tuple[int, int, int, int]:
    """Levenshtein with a backtrace, returning (S, D, I, N_ref).

    Ported verbatim from eval/controlled_eval/exp_fusion_deletions.py. `edist`
    gives the distance only; the consensus pivot and every conformance number
    need the split.
    """
    a, b = wtoks(ref), wtoks(hyp)
    n, m = len(a), len(b)
    d = [[0] * (m + 1) for _ in range(n + 1)]
    op = [[""] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        d[i][0], op[i][0] = i, "D"
    for j in range(1, m + 1):
        d[0][j], op[0][j] = j, "I"
    for i in range(1, n + 1):
        ai = a[i - 1]
        di, dim1 = d[i], d[i - 1]
        opi = op[i]
        for j in range(1, m + 1):
            if ai == b[j - 1]:
                di[j], opi[j] = dim1[j - 1], "M"
                continue
            sub, dele, ins = dim1[j - 1] + 1, dim1[j] + 1, di[j - 1] + 1
            best = min(sub, dele, ins)
            di[j] = best
            opi[j] = "S" if best == sub else ("D" if best == dele else "I")
    s = dd = ii = 0
    i, j = n, m
    while i > 0 or j > 0:
        o = op[i][j]
        if o == "M":
            i, j = i - 1, j - 1
        elif o == "S":
            s += 1
            i, j = i - 1, j - 1
        elif o == "D":
            dd += 1
            i -= 1
        else:
            ii += 1
            j -= 1
    return s, dd, ii, n


def tokenize_words(raws: list[str]) -> tuple[list[str], list[int]]:
    """Normalize a provider's raw words into one flat token stream + owner map.

    Returns `(tokens, owner)` where `owner[t]` is the index into `raws` of the
    raw word that produced `tokens[t]`. A raw word that normalizes to nothing
    (pure punctuation) contributes no tokens and never appears in `owner`; a raw
    word that normalizes to several tokens appears several times.
    """
    tokens: list[str] = []
    owner: list[int] = []
    for r, w in enumerate(raws):
        for t in wtoks(w):
            tokens.append(t)
            owner.append(r)
    return tokens, owner

File: src/lib/fusion/engine/normalize.ts
/**
 * The normalizer, ported from `fusion/normalize.py` and required to match it
 * exactly. Every frozen WER in `fusion/CONTRACT.md` was measured through the
 * Python; a token this produces differently is a word the fusion may choose
 * differently. Do not "improve" anything here.
 *
 * Two places where the languages do not line up on their own:
 *
 *  - Python's `\w` on a str is Unicode-aware. JavaScript's `\w` is ASCII only,
 *    so the class is spelled out. Measured against Python over every code point
 *    below U+3000 plus samples above it, `[\p{L}\p{N}_]` and Python's `\w`
 *    agree everywhere except U+1C89 and U+1C8A, which this Node's ICU does not
 *    yet classify as letters. Neither can occur in Greek council speech, and
 *    both runtimes' Unicode versions are recorded in the oracle index.
 *  - `unicodedata.category(c) != "Mn"` is `\p{Mn}`, not `\p{M}`: the wider class
 *    would also strip Mc and Me, which the Python keeps.
 *
 * Final sigma needs no special handling: Python's `str.lower()` and JavaScript's
 * `toLowerCase()` both apply the Final_Sigma rule, so `ΟΔΟΣ` ends in `ς` in
 * both. `ς` and `σ` stay distinct tokens, as the Python docstring insists.
 */

export const NORMALIZER_REV = "wtoks/1";

const COMBINING_MARKS = /\p{Mn}/gu;
const WORD = /[\p{L}\p{N}_]+/gu;

export function norm(s: string | null | undefined): string {
    return (s ?? "").normalize("NFD").replace(COMBINING_MARKS, "").toLowerCase();
}

export function wtoks(s: string | null | undefined): string[] {
    return norm(s).match(WORD) ?? [];
}

/** Levenshtein distance over any two sequences. */
export function edist<T>(a: readonly T[], b: readonly T[]): number {
    const n = a.length;
    const m = b.length;
    if (n === 0) return m;

    let prev = Array.from({ length: m + 1 }, (_, j) => j);
    for (let i = 1; i <= n; i++) {
        const cur = new Array<number>(m + 1);
        cur[0] = i;
        for (let j = 1; j <= m; j++) {
            cur[j] = Math.min(
                prev[j] + 1,
                cur[j - 1] + 1,
                prev[j - 1] + (a[i - 1] !== b[j - 1] ? 1 : 0),
            );
        }
        prev = cur;
    }
    return prev[m];
}

export interface Sdi {
    s: number;
    d: number;
    i: number;
    nRef: number;
}

/**
 * Levenshtein with a backtrace, returning substitutions, deletions, insertions
 * and the reference length.
 *
 * The tie-break order is load-bearing and is the Python's: when a cell can be
 * reached equally cheaply by more than one edit, substitution wins, then
 * deletion, then insertion. A port that prefers them in another order returns
 * the same distance and a different S/D/I split, and the deletion rate is the
 * number this project watches most closely.
 */
export function sdi(ref: string, hyp: string): Sdi {
    const a = wtoks(ref);
    const b = wtoks(hyp);
    const n = a.length;
    const m = b.length;

    const d: number[][] = Array.from({ length: n + 1 }, () => new Array<number>(m + 1).fill(0));
    const op: string[][] = Array.from({ length: n + 1 }, () => new Array<string>(m + 1).fill(""));

    for (let i = 1; i <= n; i++) {
        d[i][0] = i;
        op[i][0] = "D";
    }
    for (let j = 1; j <= m; j++) {
        d[0][j] = j;
        op[0][j] = "I";
    }

    for (let i = 1; i <= n; i++) {
        const ai = a[i - 1];
        for (let j = 1; j <= m; j++) {
            if (ai === b[j - 1]) {
                d[i][j] = d[i - 1][j - 1];
                op[i][j] = "M";
                continue;
            }
            const sub = d[i - 1][j - 1] + 1;
            const del = d[i - 1][j] + 1;
            const ins = d[i][j - 1] + 1;
            const best = Math.min(sub, del, ins);
            d[i][j] = best;
            op[i][j] = best === sub ? "S" : best === del ? "D" : "I";
        }
    }

    let s = 0;
    let dd = 0;
    let ii = 0;
    let i = n;
    let j = m;
    while (i > 0 || j > 0) {
        const o = op[i][j];
        if (o === "M") {
            i -= 1;
            j -= 1;
        } else if (o === "S") {
            s += 1;
            i -= 1;
            j -= 1;
        } else if (o === "D") {
            dd += 1;
            i -= 1;
        } else {
            ii += 1;
            j -= 1;
        }
    }
    return { s, d: dd, i: ii, nRef: n };
}

export interface TokenizedWords {
    tokens: string[];
    /** `owner[t]` indexes the raw word that produced `tokens[t]`. */
    owner: number[];
}

/**
 * Normalize a provider's raw words into one flat token stream plus the map back.
 *
 * A raw word that normalizes to nothing (pure punctuation) contributes no
 * tokens and never appears in `owner`. One that normalizes to several tokens
 * appears several times. The fused output has to name the raw word each chosen
 * token came from, so this map is what keeps a chosen token attributable.
 */
export function tokenizeWords(raws: readonly string[]): TokenizedWords {
    const tokens: string[] = [];
    const owner: number[] = [];
    for (let r = 0; r < raws.length; r++) {
        for (const t of wtoks(raws[r])) {
            tokens.push(t);
            owner.push(r);
        }
    }
    return { tokens, owner };
}
